core/src/component.rs

```diff
@@ -9,11 +9,46 @@ mod id;
 pub mod registry;
 
 pub use self::{handle::Handle, id::Id, registry::Registry};
-pub use abscissa_derive::Component;
+pub use abscissa_derive::{Component, Injectable};
 
 use crate::{application::Application, shutdown::Shutdown, FrameworkError, Version};
 use std::{any::Any, cmp::Ordering, fmt::Debug, slice::Iter};
 
+/// The aspect of application components associated with dependency injection.
+///
+/// This exists as a separate trait to enable separately deriving this trait and
+/// manually implementing [`Component`]. See the [`Component`] documentation for
+/// full details.
+pub trait Injectable<A>: AsAny + Debug + Send + Sync
+where
+    A: Application,
+{
+    /// Identifier for this component.
+    ///
+    /// These are the Rust path (e.g. `crate_name:foo::Foo`) by convention.
+    fn id(&self) -> Id;
```

Comment:
Question: why did `Component::id` take `&self`? The derive doesn't make use of it, and AFAICT it makes accessing a compoent via `APP.state()` rather tricky (rather than just `APP.state().components().get_by_id(MyComponent::id())`).

Reply:
Ah, and now I see you're supposed to do `APP.state().components().get_downcast_ref::<MyComponent>()`. It would be useful to document this more prominently.